### PR TITLE
create henson plugin

### PR DESCRIPTION
**Pull request by stephanenicolas**
#164 : this is the first PR to create the henson-plugin: it should work. There is a functional test, but really not so precise yet.

The plugin allows to add a new source set that will contain the models for dart&henson 3 navigation.

The navigation source set is located in 'src/navigation/java/'.
Each build type, flavor and variant can use a specific navigation source set. The source set should contain models only and will be processed by the dart & henson APs. It will expose a `navigationApi` configuration for a module using the plugin. This artifact will be a jar containing intent builders and the henson class, the extra binder, and the model class itself.

The plugin also adds automatically the navigation jar to the dependencies of the main source set.

See https://github.groupondev.com/snicolas/Henson_structure for an example

The plugin also manages the dependencies to dart, using the right annotation processors on the new navigation source set. The only things users have to do are : 
* to use the plugin
* create the directory folder `src/navigation/java` (or any variant aware equivalent such as `src/navigation/red/java` or `src/navigation/debug/java` or `src/navigation/red/debug/java`.
* create a model, annotate it
* add if needed dependencies to compile their model using configurations such as `navigationApi`, `redNavigationApi`, `debugNavigationApi`, `redDebugNavigationApi`. All android scopes are supported: Api, Implementation, compileOnly and annotationProcessor.

The plugin is roughly unit tested for now.

### CODE REVIEW

**Review by dlemures — APPROVED** [review]
🛳 it, not very familiar with groovy though...

File: henson-plugin/build.gradle [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+check.dependsOn functionalTest
+
+project.afterEvaluate {
+    tasks.getByName('compileGroovy').doFirst {
+        //we create a file in the plugin project containing the current project version
+        //it will be used later by the plugin to add the proper version of
+        //dependencies to the project that uses the plugin
+        def prop = new Properties()
+        def propFile = new File("${project.rootDir}/henson-plugin/src/main/resources/build.properties")
```
Comment (kshirsagarps) [inline henson-plugin/build.gradle]: Does the D&H version go here? If so, would be nice to use it for the composite builds needed for D&H sample.

File: henson-plugin/src/main/groovy/dart/henson/plugin/HensonPlugin.groovy [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        //we do the following for all sourcesets, of all build types, of all flavors, and all variants
+        //  create source sets
+        //  create configurations
+        //  create dependencies
+        //  create tasks: compile and jar
+        //  create artifacts
+
+        //the main source set
+        def suffix = ""
+        def pathSuffix = "main/"
```
Comment (kshirsagarps) [inline henson-plugin/src/main/groovy/dart/henson/plugin/HensonPlugin.groovy]: Can some of this be constants?

**Review by kshirsagarps — COMMENTED** [review]
I see that most of the files do not end with a new line. Can you please add a new line.
Everything else looks good.

File: henson-plugin/build.gradle (hunk 1 above)
Comment (stephanenicolas) [reply henson-plugin/build.gradle]: Yes, if there is no better way to do it.

**Review by ll0295478 — APPROVED** [review]

**Pull request merged**